File: services/golang/internal/handler/project_handler.go
package handler

import (
	"context"

	"connectrpc.com/connect"
	"github.com/google/uuid"
	"google.golang.org/protobuf/types/known/timestamppb"

	commonv1 "github.com/igorrmotta/api-corestack/services/golang/gen/common/v1"
	projectv1 "github.com/igorrmotta/api-corestack/services/golang/gen/project/v1"
	"github.com/igorrmotta/api-corestack/services/golang/gen/project/v1/projectv1connect"
	"github.com/igorrmotta/api-corestack/services/golang/internal/repository"
	"github.com/igorrmotta/api-corestack/services/golang/internal/service"
)

type ProjectHandler struct {
	projectv1connect.UnimplementedProjectServiceHandler
	svc *service.ProjectService
}

func NewProjectHandler(svc *service.ProjectService) *ProjectHandler {
	return &ProjectHandler{svc: svc}
}

func (h *ProjectHandler) CreateProject(ctx context.Context, req *connect.Request[projectv1.CreateProjectRequest]) (*connect.Response[projectv1.CreateProjectResponse], error) {
	workspaceID, err := uuid.Parse(req.Msg.WorkspaceId)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}
	p, err := h.svc.Create(ctx, repository.CreateProjectParams{
		WorkspaceID: workspaceID,
		Name:        req.Msg.Name,
		Description: req.Msg.Description,
	})
	if err != nil {
		return nil, toConnectError(err)
	}
	return connect.NewResponse(&projectv1.CreateProjectResponse{
		Project: projectToProto(p),
	}), nil
}

func (h *ProjectHandler) GetProject(ctx context.Context, req *connect.Request[projectv1.GetProjectRequest]) (*connect.Response[projectv1.GetProjectResponse], error) {
	id, err := uuid.Parse(req.Msg.Id)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}
	p, err := h.svc.GetByID(ctx, id)
	if err != nil {
		return nil, toConnectError(err)
	}
	return connect.NewResponse(&projectv1.GetProjectResponse{
		Project: projectToProto(p),
	}), nil
}

func (h *ProjectHandler) ListProjects(ctx context.Context, req *connect.Request[projectv1.ListProjectsRequest]) (*connect.Response[projectv1.ListProjectsResponse], error) {
	workspaceID, err := uuid.Parse(req.Msg.WorkspaceId)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}
	var params repository.ListProjectsParams
	params.WorkspaceID = workspaceID
	if req.Msg.Pagination != nil {
		params.PageSize = req.Msg.Pagination.PageSize
		params.PageToken = req.Msg.Pagination.PageToken
	}
	list, err := h.svc.List(ctx, params)
	if err != nil {
		return nil, toConnectError(err)
	}
	projects := make([]*projectv1.Project, len(list.Projects))
	for i, p := range list.Projects {
		projects[i] = projectToProto(&p)
	}
	return connect.NewResponse(&projectv1.ListProjectsResponse{
		Projects: projects,
		Pagination: &commonv1.PaginationResponse{
			NextPageToken: list.NextPageToken,
			TotalCount:    list.TotalCount,
		},
	}), nil
}

func (h *ProjectHandler) UpdateProject(ctx context.Context, req *connect.Request[projectv1.UpdateProjectRequest]) (*connect.Response[projectv1.UpdateProjectResponse], error) {
	id, err := uuid.Parse(req.Msg.Id)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}
	p, err := h.svc.Update(ctx, repository.UpdateProjectParams{
		ID:          id,
		Name:        req.Msg.Name,
		Description: req.Msg.Description,
		Status:      req.Msg.Status,
	})
	if err != nil {
		return nil, toConnectError(err)
	}
	return connect.NewResponse(&projectv1.UpdateProjectResponse{
		Project: projectToProto(p),
	}), nil
}

func (h *ProjectHandler) DeleteProject(ctx context.Context, req *connect.Request[projectv1.DeleteProjectRequest]) (*connect.Response[projectv1.DeleteProjectResponse], error) {
	id, err := uuid.Parse(req.Msg.Id)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}
	if err := h.svc.Delete(ctx, id); err != nil {
		return nil, toConnectError(err)
	}
	return connect.NewResponse(&projectv1.DeleteProjectResponse{}), nil
}

func projectToProto(p *repository.Project) *projectv1.Project {
	return &projectv1.Project{
		Id:          p.ID.String(),
		WorkspaceId: p.WorkspaceID.String(),
		Name:        p.Name,
		Description: p.Description,
		Status:      p.Status,
		CreatedAt:   timestamppb.New(p.CreatedAt),
		UpdatedAt:   timestamppb.New(p.UpdatedAt),
	}
}

File: services/typescript/src/handler/project-handler.ts
import type { ConnectRouter } from "@connectrpc/connect";
import { ConnectError, Code } from "@connectrpc/connect";
import { ProjectService as ProjectServiceDef } from "../../gen/project/v1/project_pb.js";
import { ProjectService } from "../service/project-service.js";
import { toConnectError } from "./error-mapping.js";
import { timestampFromDate } from "@bufbuild/protobuf/wkt";

export function registerProjectHandler(router: ConnectRouter, svc: ProjectService) {
  router.service(ProjectServiceDef, {
    async createProject(req) {
      if (!req.workspaceId) {
        throw new ConnectError("workspace_id is required", Code.InvalidArgument);
      }
      try {
        const p = await svc.create({
          workspaceId: req.workspaceId,
          name: req.name,
          description: req.description,
        });
        return {
          project: {
            id: p.id,
            workspaceId: p.workspaceId,
            name: p.name,
            description: p.description,
            status: p.status,
            createdAt: timestampFromDate(p.createdAt),
            updatedAt: timestampFromDate(p.updatedAt),
          },
        };
      } catch (err) {
        throw toConnectError(err);
      }
    },

    async getProject(req) {
      if (!req.id) {
        throw new ConnectError("id is required", Code.InvalidArgument);
      }
      try {
        const p = await svc.getById(req.id);
        return {
          project: {
            id: p.id,
            workspaceId: p.workspaceId,
            name: p.name,
            description: p.description,
            status: p.status,
            createdAt: timestampFromDate(p.createdAt),
            updatedAt: timestampFromDate(p.updatedAt),
          },
        };
      } catch (err) {
        throw toConnectError(err);
      }
    },

    async listProjects(req) {
      if (!req.workspaceId) {
        throw new ConnectError("workspace_id is required", Code.InvalidArgument);
      }
      try {
        const list = await svc.list({
          workspaceId: req.workspaceId,
          pageSize: req.pagination?.pageSize ?? 0,
          pageToken: req.pagination?.pageToken ?? "",
        });
        return {
          projects: list.projects.map((p) => ({
            id: p.id,
            workspaceId: p.workspaceId,
            name: p.name,
            description: p.description,
            status: p.status,
            createdAt: timestampFromDate(p.createdAt),
            updatedAt: timestampFromDate(p.updatedAt),
          })),
          pagination: {
            nextPageToken: list.nextPageToken,
            totalCount: list.totalCount,
          },
        };
      } catch (err) {
        throw toConnectError(err);
      }
    },

    async updateProject(req) {
      if (!req.id) {
        throw new ConnectError("id is required", Code.InvalidArgument);
      }
      try {
        const p = await svc.update({
          id: req.id,
          name: req.name,
          description: req.description,
          status: req.status,
        });
        return {
          project: {
            id: p.id,
            workspaceId: p.workspaceId,
            name: p.name,
            description: p.description,
            status: p.status,
            createdAt: timestampFromDate(p.createdAt),
            updatedAt: timestampFromDate(p.updatedAt),
          },
        };
      } catch (err) {
        throw toConnectError(err);
      }
    },

    async deleteProject(req) {
      if (!req.id) {
        throw new ConnectError("id is required", Code.InvalidArgument);
      }
      try {
        await svc.delete(req.id);
        return {};
      } catch (err) {
        throw toConnectError(err);
      }
    },
  });
}
